### ISSUE TITLE
Better handling and display of inverse gates

### ISSUE THREAD

**Issue by LOICVAL42** [ColibrITD-SAS/mpqp#103]
# Actual handling of inverses

In main's branch of MPQP, inverses are (nearly) not handled.
What I mean is that every time you want to have the inverse of a gate, it computes it **every** time, which when used multiple times can be very heavy performance-wise.
For instance, if I were to do gate reduction optimisation, I would either need to create a handler class or store the inverses of the gates in the circuit.
The other main issue I see; which may also affect regular users (i.e. using MPQP to create circuits, not developing that much stuff around it), is the displaying of adjoint gates. Indeed, for most gates it is not an issue as they're mostly hermitian. However, for non-hermitian gates which can be quite common (such as S or T gate), it gets very confusing.

# Example of confusion

When trying to apply the following optimisation pattern from [this paper](https://www.nature.com/articles/s41534-018-0072-4.pdf) (where P is the phase gate S):
![image](https://github.com/user-attachments/assets/d9554a35-3f4a-448d-ad55-6d9f7043e0b1)

With standard MPQP, we would have this:
```
Circuit:
   ┌───┐┌───┐┌───┐
q: ┤ H ├┤ S ├┤ H ├
   └───┘└───┘└───┘
Optimised circuit
   ┌───┐┌───┐┌───┐
q: ┤ S ├┤ H ├┤ S ├
   └───┘└───┘└───┘
```

Which is obviously very confusing.

# Display suggestion

The suggested display could be:

```
Circuit:
   ┌───┐┌───┐┌───┐
q: ┤ H ├┤ S ├┤ H ├
   └───┘└───┘└───┘
Optimised circuit
   ┌────┐┌───┐┌────┐
q: ┤ S† ├┤ H ├┤ S† ├
   └────┘└───┘└────┘
```
or
```
Circuit:
   ┌───┐┌───┐┌───┐
q: ┤ H ├┤ S ├┤ H ├
   └───┘└───┘└───┘
Optimised circuit
   ┌─────┐┌───┐┌─────┐
q: ┤ Sdg ├┤ H ├┤ Sdg ├
   └─────┘└───┘└─────┘
```
In case of ASCII only display.

# Proposed modification

## is_hermitian property

I suggest adding a boolean property `is_hermitian` to the `Gate` class, which can be set at `True` or `False` in the constructor but would default to None.
It would allow for users to quickly know if a matrix is hermitian or not. If `is_hermitian` is not defined, the gate's inverse would be calculated once to set it to `True` or `False`. In this case, the inverse could also be stored just-in-case of further use if the operator is not hermitian. Standard gates could also have their inverse matrices already filled, if not hermitian.

## Better labels with adjoint operators

For this one, the `is_hermitian` property would be needed to avoid costly computation for standard gates (X, Y, Z...).
In this case, the code of the `inverse()` function could be the following:

```python
def inverse(self) -> Gate:
        from mpqp.core.instruction.gates.custom_gate import CustomGate

        return CustomGate(
            # This line could use a precomputed or prestored adjoint matrix and could store it in this gate
            UnitaryMatrix(self.to_matrix().transpose().conjugate()),
            self.targets,
            # is_hermitian is a function as it might need to be computed before accessing
            self.label + "†" if not self.is_hermitian() else "",
        )
```

## Contributing

If needed, I can try making a fork and implementing those 2 features in the following week which I'd submit in a Pull Request once done.

**Comment by JulienCalistoTD** [comment]
Thank you for your comment. We have created a branch to solve this problem [fix: better handling inverse gates](https://github.com/ColibrITD-SAS/mpqp/pull/104.)

**Comment by hJaffaliColibritd** [comment]
This issue was solved in the latest release of the library (0.3.0). I close the issue.
